### ISSUE TITLE
[ADMINSTRATIVE] Miscellaneous thoughts and questions on CI/CD, DevOps, and infra

### ISSUE THREAD

**Issue by disposedtrolley** [omshub/website#7]
👋 Hey folks! I have a bunch of questions on how we'd like to set up our infrastructure and deployment process. We can try and use this issue as a dumping ground for now, and break down the work into standalone issues as the tasks become clear.

## Infrastructure

1. I've noticed that we have live pages at [omshub.org](http://omshub.org/) and [dev.omshub.org](http://dev.omshub.org/). Where are these being hosted right now? Where do we eventually want them to be hosted?
2. How and where do we want to persist our data?
3. How are we managing API keys, SSH keys, and other kinds of secrets for any infrastructure we've provisioned?

## Deployment

1. How is the NextJS frontend being deployed?
2. What kind of [rendering](https://nextjs.org/learn/foundations/how-nextjs-works/rendering) do we want for NextJS? This will inform the components that we'll need to deploy and maintain.
4. How many environments do we want? Dev and Prod?
5. What kind of workflow do we want for deploying changes? I'm a fan of continuous integration, so merges to the default branch should automatically trigger a build, test, and deploy. The PRs themselves should undergo a similar series of checks before being able to be merged.

## Monitoring

(This may be a little premature, but fun to think about!)

1. How will we know if the site is working?
2. What kind of logging and error reporting do we need? If a change breaks the site for users, where will we be alerted?

**Comment by ctran4347** [comment]
Hi @disposedtrolley, 

Hopefully I can answer your questions:

## Infrastructure
1. The current live pages for [omshub.org](omshub.org) and [dev.omshub.org](dev.omshub.org) are being hosted off of [Vercel](https://vercel.com/) right now. [omshub.org](omshub.org) for the main branch and [dev.omshub.org](dev.omshub.org) for the development branch
2. In terms of user data/reviews, it's still in discussion but it seems like folks are in favor of having the database hosted on [FireBase](https://firebase.google.com/).
3. We currently do not have an infrastructure in-place for managing secrets. We can definitely discuss this when the time comes. Currently, [Vercel](https://vercel.com/) handles all the deployment along with the secrets involved. When the time comes, you can also transfer ownership on that platform. 
## Deployment
1. The NextJS frontend is being deployed via [Vercel](https://vercel.com/). The repo was created off of a starter template and configured with Material UI. 
2. Currently, the community hasn't discussed what time of rendering we want so we can leave that up for discussion.
3. Right now, I just have to environments: Dev and Prod where the branch that deploys to Prod site is protected with reviewers. We can add as many environments as we want in the future. I believe [Vercel](https://vercel.com/) automatically creates subdomains for any branch in the repo.
4. I am definitely a fan too! Currently right now, that is what the repo is setup to do. I have it configured to build and deploy to [dev.omshub.org]( dev.omshub.org) when changes are made to dev and respectively for main ([omshub.org](omshub.org)). Only difference is that the main branch protected and requiring several reviewers.
## Monitoring 
1. That I don't know haha. I believe [Vercel](https://vercel.com/) might have something included so I would look into that more. Discussions will be made with the community.
2. In terms of logging and error, [Vercel](https://vercel.com/) has [something included](https://vercel.com/docs/concepts/deployments/logs) for the build and deploy process. It will also show during the deploy/build process on Github that it failed (red X next to recent commits)

**Comment by disposedtrolley** [comment]
Thanks @ctran4347, that’s really helpful!

It looks like Vercel is happily doing the heavy lifting for us right now. Am I correct in assuming that the branch deploys have been configured from Vercel’s end (I.e they’re monitoring our repo for changes to know when to deploy)?

**Comment by ctran4347** [comment]
Yes! Branch deploys are done on Vercel's end. We just have to specify the branch and the domain/subdomain as long as we have it in the configuration. Any changes will essentially reflect within a few minutes onto the respective sites. ([omshub.org](omshub.org) or [dev.omshub.org](dev.omshub.org))

**Comment by adamwespiser** [comment]
What's the argument for/against Firebase?
Would it be cheaper/more cost effective to either host a database, use a cloud database service, or something else?

My concern with Firebase is basically that we'll get hosed on server costs, run over the limit, then whoever owns the account will be personally on the hook. I understanding using Firebase would be easy, but we have the dev cycles, or at least the interest, to implement something a little more boring that could be cheaper in the long run.

**Comment by CBlumini** [comment]
> What's the argument for/against Firebase? Would it be cheaper/more cost effective to either host a database, use a cloud database service, or something else?
> 
> My concern with Firebase is basically that we'll get hosed on server costs, run over the limit, then whoever owns the account will be personally on the hook. I understanding using Firebase would be easy, but we have the dev cycles, or at least the interest, to implement something a little more boring that could be cheaper in the long run.

So this is me being a noob. Is the idea with Firebase that it makes it easier to host our database? So the frontend is hosted by one company and then the database is hosted by another? 

I hadn't head of it before so this was interesting to me: https://www.geeksforgeeks.org/firebase-introduction/

My point of reference right now is a Heroku site I've been working on where that houses the Dashboard I've built in dash and then the data eventually goes into a Postgre DB also hosted there. Maybe I am thinking too small when I think of what "backend" involves.

**Comment by awpala** [comment]
Among other things, firebase also streamlines things like single-sign on (e.g., via gmail account, etc.), which are otherwise more tedious to implement from scratch

**Comment by awpala** [comment]
But not strictly married to using firebase, it **_may_** be better to go with something like Linode or DigitalOcean. Definitely something we need to explore further...

**Comment by disposedtrolley** [comment]
Firebase will be much simpler than rolling our own infrastructure (but arguably less fun, and likely more expensive).

If I take a conservative guess at our scale, a $5/mo Digital Ocean or Linode box would probably do the trick. We're mostly serving static content at this point, with authentication being the next "interactive" component we build out. We can probably get quite far with a single box and a SQLite database running on the same machine for persistence.

We absolutely should not roll our own auth, but we can use that component of Firebase separately to their storage (it looks like it's [priced separately](https://firebase.google.com/pricing) too).

**Comment by driscoll42** [comment]
Can I suggest this issue also include the discussion around the servicing accounts from #8? I think it fits better here in this discussion

**Comment by awpala** [comment]
Closing issue for relevance. We have elected to move forward with Firebase as of July 2022. We will benchmark usage/cost to determine its suitability for longer-term use (and have attempted to optimize our data modeling and usage patterns accordingly to Firebase's stated cost models), however, due to lack of any substantial progress/activity on backend work as of early June 2022, there is no other feasible/suitable alternative as of this writing (i.e., this is a major blocker/bottleneck; at this point in time, it's either Firebase, or no site/project at all--in the latter case, the debate is moot either way).
